## Term Deposit Predictive ML Model

### Project Development
This project was developed locally with Visual Studio Code and GitHub version control.

Please check this project @ [GitHub page](https://caiocvelasco.github.io/) or @ [GitHub Repository](https://github.com/caiocvelasco/ml-project-bank-term-deposit/blob/4f1bfa72d17cefdfe2a515054f1f2cfffa3c7749/bank-term-deposit-notebook.ipynb).

### Problem Definition

A bank would like to understand the main factors driving customers to subscribe a Term Deposit.

### Objective
Our goal is to build a classifier (supervised machine Learning model) to predict whether customers would subscribe a Term Deposit. In case we find high accuracy in our model, it would imply that the bank marketing department could benefit from an efficient use of available resources by focusing on the "right" customers (given by our _classifier_). Moreover, we want to understand which other factors might also affect customers' decision.

* Recall: Classification Model

    The goal of a classification model is to build a data-driven model that learns an unknown underlying function that maps several input variables, which characterize an item (e.g., bank client, with one labeled output target (e.g., type of bank deposit sell: “failure” or “success”). 
    _[Moro et al., 2014] S. Moro, P. Cortez and P. Rita. A Data-Driven Approach to Predict the Success of Bank Telemarketing. Decision Support Systems, Elsevier, 62:22-31, June 2014_

### Background and Terminology

Suppose a bank is interested in predicting whether its customers would subscribe a _Bank Term Deposit_. 

**What is a term deposit and why does someone want to invest their money?**

Before answering these questions, let's look into some important definitions and then into some insights from Economic Theory. This will help understand the scenario.

**Financial Instruments**

A financial instrument is an asset that can be traded between entities. An agreement (or contract) will hold both parties accountable for such transaction.
We can divide financial instruments into different classes of financial instruments. An example is the _Debt-Based Financial Instruments_.

**Debt-Based Financial Instruments**

The term _debt_ comes from the fact that some financial institutions want to raise capital by borrowing money from people. When this transaction happens, the financial institution has the obligation to pay back at a later date, with interest. Thus, the financial institution has a _debt_ with the lender.
Such instruments can also be interpreted as a _Certificate of Deposit_, i.e., you _deposit_ money into an account at a financial institution for some period of time in exchange for future compensation.

**Maturity**

Maturity is the period of time it takes for the lender to withdraw this or her money from the bank. Maturity date is the agreed-upon date on which the investment ends.

**Fixed-term Investment**

When a financial instrument has a fixed time period of investment, the maturity is known at the time of issuance and the lender cannot withdraw its money befeore maturity without incurring on a _early withdrawal penalty_. This penalty is important to the bank, because a bank makes money by lending its customers deposits to other entities and profiting on top of that. Banks need to reduce risk in their operations and such penalties account for the unexpected costs created by customers who decide to withdraw their deposits before maturity.  

**Term Deposit**

Term Deposit is an example of a _Fixed-Term, Debt-Based Financial Instrument_, also called a _Certificate of Deposit_.
Moreover, a Term Deposit is a risk-free investment since they are backed by the government.
In sum:
* When a customer subscribes to a **Term Deposit** he or she will deposit (or invest) money into a bank account, agreeing not to withdraw their funds for a fixed period in return for future compensation (interest), i.e., before maturity.

**The Decision Problem faced by an Investor**

An investor is an individual (or entity) who makes certain decisions regarding how he or she spends his or her money.
* How much of my money should I use in consumption (goods or services) now?
* From what is left, how much should I spend on investment opportunities?

Such decisions are unscapable to the individual because resourses are limited. You only have a certain amount of money and you need to make decisions regarding how to spend it.
We say that the investor faces a _trade-off_ between consuming less now to (hopefully) consume more in the future, i.e.:
* Spend less in consumption at the beginning of a period in exchange for an investment opportunity that will hopefully earn profits at the end of the period.

**Economic Incentives behind Consumer Behavior & Investment Opportunities**

An _Economic Incentive_ is something that motivates or drives one to behave in a certain way. The most simple example in Economics is: Prices.\
_Prices_ provide an incentive mechanism that might induce people's behavior in a certain way. For example, if the price of apples go up, you will probably buy less apples (and buy more of some other good or service).

Good points to notice:
* In periods of rising interest rates in society (usually started by the Central Bank), consumers are more likely to invest because it is better to lend money to the bank than to spend it on consumption now. Thus, when the interest rate is higher, people tend to invest more, hence changing their status-quo behavior. Thus, future expectations of _higher interest rates_ could help predict consumer behavior towards subscribing to **Term Deposits**.
* **Term deposits** do not keep up with inflation. If the rate on a term deposit is 3% and the inflation rate. is 3.5%, the customer is not earning enough to compensate for price increases in the economy. Thus, _Inflation Rate_ maybe a strong underlying _economic incentive_ that could help predict consumer behavior towards subscribing to **Term Deposits**.

### Data Collection
Data was already available on a _csv_ format. So I'll skip this section.

### Importing Libraries

In [1]:
# !pip install seaborn pandas matplotlib numpy
import pandas as pd              # for data analysis
import numpy as np               # for scientific computing
import os                        # for file interactions in the user's operating system

import matplotlib.pyplot as plt  # for data visualization
import matplotlib as mpl
import seaborn as sns            # for data visualization

### Importing Dateset

1. [_Note_] It is important to import only the data we need to save up time and space. However, this dataset already contains the features we need, so I will skip that.
2. [_Importing_] The csv dataset will be imported as a dataframe using pandas.
3. [_Independence_] This notebook can be downloaded and saved anywhere in your computer (one needs only to change the name of the folder where the CVS files are located).
4. [_Automation_] Any dataset saved in the designated folder will be stored automatically.
5. [Stacking] If there is a need to combine (or stack) multiple CSV files, please read the _extra_ below.

Extra
* If there is a need to stack (or combine) multiple CSV files based on a matching _filename_ pattern, please check one of my [repositories](https://github.com/caiocvelasco/stack-combine-multiple-csv-files/blob/7854a473046e4dc5876bcf8caa69b4b3d16dc7e1/stack-combine-multiple-csv-files.ipynb). There, you'll find a code to perform this task automatically under the following assumption: the stored CSV files have the same features (aka same columns).

In [2]:
# Basic Settings
csv_folder_name = "bank_marketing"  # please, change the folder name where the CSV files are stored **if need be**
notebook_location = "C:\\Users\\caiov\\OneDrive - UCLA IT Services\\Documentos\\DataScience\\Datasets" # set the location where this notebook is saved
csv_folder_path = notebook_location + "\\" + csv_folder_name  # set path for the CSV files
os.chdir(csv_folder_path)                                     # set location of CSV files

# Save cvs data on a pandas dataframe
df = pd.read_csv("bank-full-case.csv", sep = ";", skipinitialspace = True) #skip space after delimiter

# Save a copy of the dataframe
data = df.copy()

# Basic Formatting
data.name = "Bank Term Deposit Dataset" # rename the dataset 
cols = data.columns;                    # create a list with features names

# Quick overview of a sample from the data
data.sample(5)

age         job   marital  education default  balance housing loan  \
44392   48   housemaid    single    primary      no       46      no   no   
19882   34  technician   married  secondary      no      134      no   no   
39654   24  technician    single  secondary      no     1234      no   no   
2112    31      admin.   married  secondary      no      184     yes   no   
32045   36  management  divorced   tertiary      no      102     yes   no   

        contact  day month  duration  campaign  pdays  previous poutcome    y  
44392  cellular    3   aug       113         2    180         1  failure  yes  
19882  cellular    8   aug       104         4     -1         0  unknown   no  
39654  cellular   27   may       480         2     -1         0  unknown   no  
2112    unknown   12   may       378         3     -1         0  unknown   no  
32045  cellular   14   apr       114         1     -1         0  unknown   no

### Exploratory Data Analysis - EDA

We have the data. The next step is to explore it. We'll look at:
    - Checking Data Types
    - Handling Missing Values
    - Converting Units (if needed)
    - Summarizing Continuous Variables (making sense of the data / summary statistics / outliers detection and treatment)
    - Summarizing Categorical Variables (Histograms)
    - Histogram - for the target we want to predict
    - Probability Density Function per marital status, education, and type of contact 
    - Multiple Boxplots per marital status, education, and type of contact 
    - Correlation Matrix for multiple features

* Shape and Size

In [3]:
# Check Shape and Size
print("Dataset shape: " + str(data.shape[0]) + " observations and " + str(data.shape[1]) + " features.")
print("This dataset has a total of: " + str(data.size) + " entries.")

Dataset shape: 45211 observations and 17 features.
This dataset has a total of: 768587 entries.


* Data Types

In [4]:
# Check data types
print(data.name)
print("--------------------------------------")

data.info() # get summary of dataset types
print("--------------------------------------")

Bank Term Deposit Dataset
--------------------------------------
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45211 entries, 0 to 45210
Data columns (total 17 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   age        45211 non-null  int64 
 1   job        45211 non-null  object
 2   marital    45211 non-null  object
 3   education  45211 non-null  object
 4   default    45211 non-null  object
 5   balance    45211 non-null  int64 
 6   housing    45211 non-null  object
 7   loan       45211 non-null  object
 8   contact    45211 non-null  object
 9   day        45211 non-null  int64 
 10  month      45211 non-null  object
 11  duration   45211 non-null  int64 
 12  campaign   45211 non-null  int64 
 13  pdays      45211 non-null  int64 
 14  previous   45211 non-null  int64 
 15  poutcome   45211 non-null  object
 16  y          45211 non-null  object
dtypes: int64(7), object(10)
memory usage: 5.9+ MB
----------------------------------

All features have the expected type. Notice that all non-null counts are the same, so the dataset does not seem to have missing values of the form N/A. However, let's take a closer look into it.

* A closer look into categorical variables (unique values)

In [5]:
# Check for unique values into caterogical variables

def unique_val(data):                                          # define a function (output: unique values in the categorical variables)
    for i in cols:                                             # cols is the list of features from this dataset defined in the "Importing the Dataset" section above
        if data[i].dtype == 'O':                               # check whether features are categorical variables. 'O' stand for object type.
            print('Unique values in',i,'are',data[i].unique()) # calls function unique() to find get unique values
            print('----------------------------------------------------------------------------------------------------')
unique_val(data)                  # calls the function

print("Columns with missing values set as 'unknown':")
cols[                       # uses the list of columns 'cols' and subsets it to the condition we want
    (df == 'unknown')       # makes a boolean with True/False 
    .any(axis = 0)          # returns false unless there is at least one element metting the condition withing a series or along a df axis
].to_list()                 # return a list of the values

Unique values in job are ['management' 'technician' 'entrepreneur' 'blue-collar' 'unknown'
 'retired' 'admin.' 'services' 'self-employed' 'unemployed' 'housemaid'
 'student']
----------------------------------------------------------------------------------------------------
Unique values in marital are ['married' 'single' 'divorced']
----------------------------------------------------------------------------------------------------
Unique values in education are ['tertiary' 'secondary' 'unknown' 'primary']
----------------------------------------------------------------------------------------------------
Unique values in default are ['no' 'yes']
----------------------------------------------------------------------------------------------------
Unique values in housing are ['yes' 'no']
----------------------------------------------------------------------------------------------------
Unique values in loan are ['no' 'yes']
------------------------------------------------------------

['job', 'education', 'contact', 'poutcome']

All features look good and with the expected categories for each feature. Moreover, there are no unexpected entries across categories.
However, we do see some 'unknown' values in the data for the following categories: _job, education, contact, and poutcome_. This can be interpreted as missing values and need to be trated accordingly.

* Handling Missing Values

In [6]:
print('Checking for missing values in the', data.name)
print('------------------------------------------------------------')
print(data.isnull().sum())
print('------------------------------------------------------------')

Checking for missing values in the Bank Term Deposit Dataset
------------------------------------------------------------
age          0
job          0
marital      0
education    0
default      0
balance      0
housing      0
loan         0
contact      0
day          0
month        0
duration     0
campaign     0
pdays        0
previous     0
poutcome     0
y            0
dtype: int64
------------------------------------------------------------


As we noticed, the standard missing value function cannot help us visualize missing data because missing values does not come as N/A, but as 'unknown'. Thus, we have to find ways to deal with the 'unknown' values.

Usually, one of the following is performed to handle missing values:
    - Dropping observations or features.
    - Imputation via values: filling missing data with some statistic (e.g.: mean, median) or some value (e.g.: 'unknown').
    - Imputation via prediction model.

In this dataset, there is no missing values in the N/A format. However, as we just saw above, some missing values are treated as 'unknown' values.
It is important to check whether missing information (in this case, the 'unknown') has any statistically relevant association with the outcome we want to predict. 

However, given the timeframe, I will skip this part.

* Converting Units

There is no need to convert units in the given features.

* Summarizing Continuous Variables

In this section we need to check wether the most important statistics makes sence. We'll check the max and min values. Moreover, we'll look into the summary statistics and check for outliers.

In [7]:
# Summary Statistics
data.describe()

age        balance           day      duration      campaign  \
count  45211.000000   45211.000000  45211.000000  45211.000000  45211.000000   
mean      40.936210    1362.272058     15.806419    258.163080      2.763841   
std       10.618762    3044.765829      8.322476    257.527812      3.098021   
min       18.000000   -8019.000000      1.000000      0.000000      1.000000   
25%       33.000000      72.000000      8.000000    103.000000      1.000000   
50%       39.000000     448.000000     16.000000    180.000000      2.000000   
75%       48.000000    1428.000000     21.000000    319.000000      3.000000   
max       95.000000  102127.000000     31.000000   4918.000000     63.000000   

              pdays      previous  
count  45211.000000  45211.000000  
mean      40.197828      0.580323  
std      100.128746      2.303441  
min       -1.000000      0.000000  
25%       -1.000000      0.000000  
50%       -1.000000      0.000000  
75%       -1.000000      0.000000  
max      871.000000    275.000000

Quick overview:

- [Age] Note that the average age of people in this dataset is roughly 41 years old. The median is 39 years old, which means that half of the people with an open account at this bank is less than 39 and half is older. This makes sense, since we can find both older and younger people with an open banck account. Moreover, the percentile 75th is 48 years old, which means that 75% of the people (the majority) are less than 48 years old. Both the min and max values do make sense, because there can be someone very old (95 years old) and someone just starting their adult life (18 years old).
- [Balance] Note that 75% of the people have less than 1428 euros in their bank account by the end of the year, on average. Both the min and max values do make sense, because there can be someone very rich (100k euros) and someone with a big debt (-8k euros).
- [Day] Everything looks normal for this variable. Note that half of the people we contact last in the first half of the month and the other half in the last half of the month. This could be an interesting feature because maybe some people are more prone to subscribing a term deposit when they are done paying the majority of their household bills.
- [Campaign] Note that the average number of contacts performed during this campaign for a specific client is roughly 3 times. However, we do notice that the max number does not make much sense because it would be very strange that someone was contacted 63 times during a campaign. This makes even more sence when 75% of the people were contacted at most 3 times. _We need to treat that as an outlier._
- [Pdays] Note that the majority of pepole (75% of them) were not previously contacted. For those 25% who are left, we see an average of 40 days qfter the client was last contacted from a previous campaign, with a large variation. This variation can make sense since we are looking into a small portion of the data. Someone was last contacted 871 days after the last campaign. This could also be an outlier.
- [Previous] Note that before this campaign, the majority of people were not contacted. From what is left, less than 1 contact was performed on average per client. Someone was contacted 275 times before this campaign and this seems odd. This could also be an outlier.

In [ ]:
# Boxplots
cont_data = data.select_dtypes(include='number')
cont_data.head()

* Summarizing Continuous Variables - Outliers

Outliers can be found due to varation in the data (i.e., due to variance in some features) or due to mistakes. Both cases can be dealt with. A simple way to start is to vizualize the data with boxplots.

* Summary and Outliers - Categorical Variables

In [31]:
# Category Distribution (Relative Frequency Distribution)